### ISSUE TITLE
Remove BuildTargetKind ?

### ISSUE THREAD

**Issue by smarter** [scalacenter/bsp#30]
It's currently defined as:
```scala
object BuildTargetKind {
  final val Library = 1
  final val Test = 2
  final val App = 3
  final val IntegrationTest = 4
  final val Bench = 5
}
```

Most build tools do not require the user to classify their projects in such a way, so the best a build tool could do is try to guess using heuristics, but that can also be done by the IDE itself. Moreover, this information is only needed to display something nice to the user, not to compute anything, so there really isn't any reason to keep it in the protocol in my opinion.

**Comment by jvican** [comment]
Target kinds are useful for clients that want to highlight the nature of a project; for example, an IDE that knows how to run benchmarks can use the `Bench` target kind to auto-detect benchmarks and allow users to run them from the IDE.
  
Most of the build tools (sbt, bazel, pants, maven, gradle) have the ability to define targets that are either binaries, library, tests, benchmarks or integration tests; though they do it in different ways.
  
In Bazel and pants, users define the purpose of a target by choosing the right directive (`scala_binary`, `scala_library`, etc). In sbt, maven and gradle, the definition is implicit via the use of Maven configurations (for example, in sbt you add the `IntegrationTest` configuration to a project that defines integration tests, and likewise for `Bench`). If a build tool does not know the nature of a build target, it can safely use `App` as the default.
  
All in all, the build tool picks the target kind not by heuristics, but by information given by the user.This information cannot be derived from the IDE and it's therefore required in the project.

See discussion in https://github.com/scalacenter/bsp/issues/12, https://github.com/scalacenter/bsp/commit/80e45b4981fedb86f063b9b07406f3154adc0896 and https://github.com/scalacenter/bsp/pull/5.

**Comment by smarter** [comment]
> an IDE that knows how to run benchmarks can use the Bench target kind to auto-detect benchmarks and allow users to run them from the IDE.

I don't understand this, every benchmark framework will have different ways of running them, knowing that something is a benchmark project doesn't give enough information.

> Most of the build tools (sbt, bazel, pants, maven, gradle) have the ability to define targets that are either binaries, library, tests, benchmarks or integration tests

How do you define benchmarks and integration tests in bazel or pants for example?

> in sbt you add the IntegrationTest configuration to a project that defines integration tests, and likewise for Bench

I don't see a Bench configuration in sbt.

>  If a build tool does not know the nature of a build target, it can safely use App as the default.

That doesn't seem right, for me an app is something that can be run, a random build target doesn't always have a way to be run, but it probably has some kind of API so "Library" would be a better default, in any case this would need to be documented.

> This information cannot be derived from the IDE and it's therefore required in the project.

I still don't see what useful thing the IDE is supposed to do with this information, `BuildTargetCapabilities` already has `canCompile/canRun/canTest` which are easy to give semantics to (though I don't see the point of `canCompile`, in what case do you have a build project that you cannot build?), `Test` is redundant with `canTest`, `App` is redundant with `canRun`, and `IntegrationTest` and `Bench` are too vague to be useful (and probably not something most build tools know about a project anyway).

**Comment by olafurpg** [comment]
@smarter IDEs can use the build target kind to display different icons for different targets. IntelliJ does this for example with blue and green colors for directories

![screen shot 2018-08-07 at 12 01 53](https://user-images.githubusercontent.com/1408093/43769328-c430a7e6-9a39-11e8-958f-4f83d4cba07b.png)

Another example application is a UI to run/build/test/package like https://majestic.debuggable.io/

In sbt it's possible to automatically label the "jmh" configuration as benchmarks. Build tools like cargo provide benchmarking out of the box https://doc.rust-lang.org/1.7.0/book/benchmark-tests.html and I wish more build tools would do the same.

I proposed in https://github.com/scalacenter/bsp/issues/12#issuecomment-394963232 to use `labels: List[String]` to solve the same problem and that approach has the benefit of 1) allowing a server to leave labels empty and 2) introduce new kinds without changing the protocol. Do you think that solution would be an improvement?

**Comment by jvican** [comment]
> I don't understand this, every benchmark framework will have different ways of running them, knowing that something is a benchmark project doesn't give enough information.
  
If I'm IntelliJ and want to support Jmh in the JVM world *in a build-tool agnostic way*, knowing that a target is a benchmark simplifies a lot my life! All I need to do is to find the JMH version used by inspecting the classpath of that target and do some class file analysis on its classpath (after compilation) to find the benchmarks that the user defined. With that information, I can already present the user with a UI to select the benchmark to run and tweak the usual JMH parameters.

**Comment by olafurpg** [comment]
For me, the problem with the current design is that build servers are forced to provide a `BuildTargetKind` when the build server might want to return zero or multiple target kinds. I would be open for a discussion on how to improve that situation.

**Comment by olafurpg** [comment]
Seems like the agreement is to keep `BuildTargetKind` unchanged, which means we can close this ticket. Any objections?

**Comment by jastice** [comment]
From discussion, we agreed to keep `BuildTargetKind`, but make it more flexible:

* `kind` is a string, not an int enum
* we define a set of standard `kind`s
* servers may specify their own more specific `kind` strings

**Comment by olafurpg** [comment]
@smarter the discussed agreement was to remove `kind: Int` and a field `labels: List[String]`

```diff
trait BuildTarget {
+ def labels: List[String]
}
```

The labels are free-form and can be used in any way between the client/server to communicate additional categorization of build targets. The server can leave labels empty if it wants.

I agree that clients should use the `BuildTargetCapabilities` to know whether a target can be tested or know.

**Comment by smarter** [comment]
> The labels are free-form and can be used in any way between the client/server to communicate additional categorization of build targets.

The Language Server Protocol already has a standardized way to announce extra capabilities, it's called dynamic registration: https://microsoft.github.io/language-server-protocol/specification#client_registerCapability I would be very wary of adding another under-specified mechanism, especially without usecases, it defeats the point of having a standard.

**Comment by olafurpg** [comment]
I am personally fine with removing the build target kind and not add labels. @jastice ?

**Comment by olafurpg** [comment]
Removing labels + kind would require build servers to properly document build target capabilities, currently bloop does
```scala
          val capabilities = bsp.BuildTargetCapabilities(
            canCompile = true,
            canTest = true,
            canRun = true
          )
```
cc/ @jvican

**Comment by jastice** [comment]
I think using just the capabilities `canTest` would be enough for IntelliJ, insofar that build targets are generally declared with either compile or test capability in mind, or separated by the tool for the purpose of the protocol.

Agreed something like `client/registerCapability` seems like a better fit than labels for adding anything more specific or special-cased.

**Comment by olafurpg** [comment]
Sounds good, I will follow up with removing kind with no alternative 👍

**Comment by jvican** [comment]
> if my IDE is smart enough to detect that something is a JMH project, then it can do that regardless of whether the project is identified as "Bench" or not
  
Sometimes knowing for sure that the build tool has defined that project as a benchmark can be critical for both correctness (no false positives) and performance (when the discovery of certain properties is just too expensive). Build tools like bazel and pants, even sbt, allow you to "tag" certain projects with semantics that are important for IDEs/editors.

I still believe that tags (or build target kinds) have a lot more uses on the client side rather than just `canCompile`, `canTest` and `canRun` that can be understood by the client and supported by the specification without the need of bloating the protocol with dynamic registrations or specific support for them.
  
I'm not convinced this is the right move and I wouldn't be surprised if in several months from now clients need tags to have precise and reliable information about targets that they can use to enrich the user experience or to support concrete frameworks of targets.

**Comment by olafurpg** [comment]
If the need for tags / labels arises in the future we can add this functionality. Once we have a clear use case it will be better to design the correct contract.

**Comment by smarter** [comment]
I agree with Olaf. I strongly believe that the protocol design should be driven by usage, adding things to a protocol is much easier than removing them, so it makes sense to be conservative and only add things when the pain of not having it is being felt.

**Comment by jvican** [comment]
We have current use cases, supporting specific benchmark frameworks or web frameworks is one. If IntelliJ knows a project is a benchmark, it can try to discover JMH benchmarks and run them from the IDE. If IntelliJ knows a project is a web server, it can also have specific web frameworks support (like it has now for Spring or Play) to toggle actions in the editor like "Profile web server".
  
I do not believe we should make this decision based on usage. This is a property fundamental to a project. We should not conflate tags with capabilities. They are different and tags give a lot more flexibility to clients to support the previous use cases reliably.

**Comment by olafurpg** [comment]
I am not convinced that `BuildTargetKind` or `labels: List[String]` is sufficient data to implement interfaces such as "Run benchmark" or "Profile web server". If the protocol is supposed to enable such use-cases, then I believe we will need sit together and design custom endpoints like `buildTarget/benchmark` or `buildTarget/profile`, with accompanying `BuildTargetCapabilities`.

I agree with @smarter that it it is easier to remove than add functionality, especially as the protocol matures. Given it's still early days I lean towards removing `BuildTargetKind` and don't add labels/tags until we have a better picture of how those fields will be used (I don't have any use-case for them in metals at the moment, and it seems intellij-scala's usage can be replaced with `BuildTargetCapabilities.{canTest,canCompile}`).

**Comment by jvican** [comment]
I'm fine with that, so long as we acknowledge these are important entry points we should support in the protocol sooner or later. Labels could have been a way to unblock clients to implement these behaviors without the expensive process of adding them to the protocol explicitly and then updating clients/servers.

**Comment by olafurpg** [comment]
Sounds good. Clients and servers should in theory never be blocked by the protocol because they can still agree on implementing custom endpoints on each side.

**Comment by jastice** [comment]
The IntelliJ-specific problem I have is that I want need to map targets that share the same base directory to a single IntelliJ module. If there is only one target with `canTest==true` that is fine. But multiple testing targets (eg `test` and `integrationTest`) need to be separate somehow, or at least I need to be able to select only the main targets for import and ignore the others. Neither `BuildTargetKind` nor `tags` seem like an optimal fit for this purpose, but either are workable. IS there a better way out?

**Comment by smarter** [comment]
I think a target should have a base directory as one of its attribute

**Comment by jastice** [comment]
> I think a target should have a base directory as one of its attribute

currently I assume the path part of the buildtargetid uri is the base directory. But having a separate baseDirectory uri does not solve the problem I described. In IntelliJ, sources under the same base directory can only be in one of a fixed set of categories: source, test, resource, test resource or excluded. Library dependencies can only be compile, test, runtime, provided. Basically if a target defines something that doesn't map to these concepts I need to ignore it. If I don't have information on these scopes, differing targets will be lumped together in the some scope. That will probably cause user confusion.

**Comment by olafurpg** [comment]
Here's a proposal that might address your needs @jastice 

- Use `BuildTargetCapabilities.canTest: Boolean` to determine if a target is for test or not.
- We update `SourceKind` in https://github.com/scalacenter/bsp/issues/21#issuecomment-432948810 to have three categories
```scala
object SourceItemKind {
  /** Source code that is manually edited by the user. */
  val Source = 1
  /** Data files that are manually managed by the user. */
  val Resource = 2
  /** Automatically generated source by the build and is not
    * intended to be manually edited by the user. */
  val Excluded = 3
}
```

> currently I assume the path part of the buildtargetid uri is the base directory

I think we should consider adding `BuildTarget.baseDirectory: Uri` to avoid encoding further information in the target URI. It should be an opaque implementation detail in my opinion that accidentally happens to work so far with the bloop server.

**Comment by andyscott** [comment]
@olafurpg The Bazel plugin for IntelliJ treats the whole Bazel workspace as a single module. From there, you have IntelliJ configurations for running/testing Bazel targets.

**Comment by jastice** [comment]
> Here's a proposal that might address your needs @jastice

I don't see how this would work to distinguish between a test and integrationtest config under the same base. The issue here is that these "configs" in sbt parlance have both different sets of sources and different sets of dependencies.

> I think we should consider adding `BuildTarget.baseDirectory: Uri` to avoid encoding further information in the target URI. It should be an opaque implementation detail in my opinion that accidentally happens to work so far with the bloop server.

Somewhat agree, this was confusing for me at first.

**Comment by smarter** [comment]
> I think we should consider adding BuildTarget.baseDirectory: Uri to avoid encoding further information in the target URI. It should be an opaque implementation detail in my opinion that accidentally happens to work so far with the bloop server.

100% agree

> But multiple testing targets (eg test and integrationTest) need to be separate somehow, or at least I need to be able to select only the main targets for import and ignore the others

@jastice This seems like a separate issue that should be discussed elsewhere. In any case, I'm not sure I get what we could possibly do in the protocol: if there are indeed two targets with the same base directory that are both tests, or both non-tests, then IntelliJ is free to lump them together in its UI if it cannot handle them in any other way. Also sbt uses separate source directories for the Test and IntegrationTest configuration so shouldn't that be ok ?

**Comment by stuhood** [comment]
@olafurpg : I will defer to @wisechengyi in that regard.

**Comment by jvican** [comment]
> It should be an opaque implementation detail in my opinion that accidentally happens to work so far with the bloop server.
  
Why does it happen to work accidentally in bloop?
  
> I think we should consider adding BuildTarget.baseDirectory: Uri to avoid encoding further information in the target URI.
  
What are the objective benefits of this approach?
  
I like the current encoding where we have a URI for the base directory in the build target directory ID + a query component specifying the id name. I think it's easy to reason about that the ID of a build is defined in terms of these two parameters which are explained in the specification.

**Comment by wisechengyi** [comment]
@olafurpg @stuhood assuming this was the question:
> any insights on how pants/bazel targets map to IntelliJ modules? I am personally fine with adding baseDirectory: Uri to build targets but I'd like to be sure that this does (edit) not impose unintended complications for build tools that support multiple targets within the same directory.

**Status quo**: Pants tries to map target to intellij module one to one. However, if a directory contains multiple targets, intellij-pants-plugin is packing all them into one module, thus causing unintended behaviors in IntelliJ sometimes. E.g. a resources target and a library target live in the same dir.

Is `BuildTarget.baseDirectory` supposed to be [source root](https://www.pantsbuild.org/setup_repo.html#source-roots)? If that's the case, I can imagine it being useful since Pants does export this info to IntelliJ in order to configure the module correctly.

**Comment by jvican** [comment]
> If another build server uses a different URI encodings for build targets then that would break the IntelliJ Scala client because it infers information encoded in Bloop URIs.

Oh, then it accidentally works in IntelliJ, not bloop :wink:
   
> If we explicitly document the BuildTarget.baseDirectory requirement in the specification then build servers can use any URI-encoding they like for build targets.
  
Since we're going to have the project root base directory in the bsp connection details, I would prefer that we change the protocol to guarantee that servers use the per-project base directory in the `BuildTargetIdentifier`.
  
I would then discuss whether we should keep having the project name in the URI and specify it in the protocol or adding a concrete field to it.

**Comment by jvican** [comment]
> Status quo: Pants tries to map target to intellij module one to one. However, if a directory contains multiple targets, intellij-pants-plugin is packing all them into one module, thus causing unintended behaviors in IntelliJ sometimes. E.g. a resources target and a library target live in the same dir.
  
This logic will then needed to be done on the IntelliJ side /cc @jastice The protocol will allow you to express this as independent targets. I know there are some efforts on the IntelliJ side to update their project model to work around all of these issues, Justin can talk more about that.

**Comment by smarter** [comment]
> Since we're going to have the project root base directory in the bsp connection details, I would prefer that we change the protocol to guarantee that servers use the per-project base directory in the BuildTargetIdentifier.

I dont think that can work. BuildTargetIdentifier have to be unique by definition, but two projects can share the same baseDirectory (e.g. when cross-compiling).

**Comment by jvican** [comment]
> I dont think that can work. BuildTargetIdentifier have to be unique by definition, but two projects can share the same baseDirectory (e.g. when cross-compiling).
  
Yup, that's why I follow up in the same comment saying that we should also having the project name in the URI to disambiguate projects that share the same base directory:
  
> I would then discuss whether we should keep having the project name in the URI and specify it in the protocol or adding a concrete field to it.

**Comment by smarter** [comment]
That means that now you need to specify a format for URIs that every client and every server needs to know how to parse. Why do that when we can just use fields in JSON messages ? Plus, if the BuildTargetIdentifier areom opaque, servers can use whatever representation suits them better without having to maintain any mapping table. For example in my prototype bsp implementation for mill, the BuildTargetIdentifier was just the string you pass to the mill command-line to run this target, that was really convenient.

**Comment by jvican** [comment]
I didn't marry myself to a solution: I claimed that we could either add a `name: String` to the build target or have it in the URI and specify it. I prefer that the build target identifier has the URI in its ID, just as any other thing in the protocol has (text document, source items, resource items, etc).
  
I can understand how it's simpler for your use case, though.

**Comment by olafurpg** [comment]
I opened #50 about adding `BuildTarget.baseDirectory`

@wisechengyi answered in https://github.com/scalacenter/bsp/issues/50#issuecomment-433354334 😊 

@andyscott that is an interesting approach! Judging by https://ij.bazel.build/docs/project-views.html, it seems the IntelliJ Bazel plugin doesn't expose the full bazel workspace but only the relevant targets for a single "IDE import".

For the record, the IntelliJ Bazel plugin has [`tags: List[String]`](https://ij.bazel.build/docs/tags.html) for build targets but apparently only uses it for `no-ide` to exclude targets from the IDE import, a different use-case than discussed in this issues.

**Comment by olafurpg** [comment]
The question remains what to do with `BuildTargetKind`. My understanding of the problem is that IntelliJ Scala needs a way to pair main/test targets when there are multiple (>2) targets in the same base directory. Keeping `BuildTargetKind.{Library,Application,Test,IntegrationTest,Benchmark}` doesn't fully solve this problem, but at least helps a bit with determining which targets to pair together.

There is no open proposal to solve this problem 100% so I think it's best that we keep `BuildTargetKind` unchanged and delay improvements for the next iteration of the spec.

@smarter A build server that is unsure how to populate `BuildTargetKind` can use the following logic

```scala
val kind =
  if (buildTarget.capabilities.canTest) BuildTargetKind.Test
  else if (buildTarget.capabilities.canRun) BuildTargetKind.Application
  else BuildTargetKind.Library
```

**Comment by jvican** [comment]
> My understanding of the problem is that IntelliJ Scala needs a way to pair main/test targets when there are multiple (>2) targets in the same base directory
  
Yes, this is what I wrote in the commit message [that introduced build target kind](
https://github.com/scalacenter/bsp/pull/5/commits/041be33f171d16930d84db395ed82659dbd5102a):
  
> Even though `IntegrationTest` and `Bench` are not supported by, to the
best of my knowledge, any build tool and IDE, it is still beneficial to
have the kind in the build target.
>
>For example, IntelliJ doesn't support `IntegrationTest`, but if IntelliJ
users want to see the tests that they can run in their projects,
IntelliJ has to show the test main class from both the `Test` target
kind and `IntegrationTest` target kind.
>
>Another use case where this is useful is in showing all the imported
targets, no matter if there's support for them or not, in a list of the
UI so that the user can manually trigger a bsp action on them.
  
> There is no open proposal to solve this problem 100% so I think it's best that we keep BuildTargetKind unchanged and delay improvements for the next iteration of the spec.
  
Wouldn't the solution we all agreed on and that you proposed about `tags: List[String]` solve this problem? Free-form strings give freedom to both the client and the server.

**Comment by jastice** [comment]
> Wouldn't the solution we all agreed on and that you proposed about tags: List[String] solve this problem? Free-form strings give freedom to both the client and the server.

For my purposes either "kind" or "tags" will work, with tags being a bit more flexible. The "no-ide" tag for instance is a useful feature already supported for other build tools one way or another.
